### PR TITLE
Handle references to deep schema definitions

### LINKED ISSUE #553: References to definitions deeper than immediate children of `#/definitions` are not supported
### Prerequisites

- [x] I have read the [documentation](https://github.com/mozilla-services/react-jsonschema-form/blob/master/README.md#readme);
- [x] In the case of a bug report, I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.

### Description

References to definitions deeper than immediate children of `#/definitions` are not supported.

### Steps to Reproduce

1. Define a schema that references definitions deeper than immediate children of `#/definitions`:
    ```json
    {
      "definitions": {
        "testdef": {
          "type": "object",
          "properties": {
            "bar": {
              "type": "string"
            }
          }
        }
      },
      "type": "object",
      "properties": {
        "foo": {
          "$ref": "#/definitions/testdef/properties/bar"
        }
      }
    }
    ```
2. Attempt to create a form using that schema ([shared playground link](https://mozilla-services.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6e30sInNjaGVtYSI6eyJkZWZpbml0aW9ucyI6eyJ0ZXN0ZGVmIjp7InR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7ImJhciI6eyJ0eXBlIjoic3RyaW5nIn19fX0sInR5cGUiOiJvYmplY3QiLCJwcm9wZXJ0aWVzIjp7ImZvbyI6eyIkcmVmIjoiIy9kZWZpbml0aW9ucy90ZXN0ZGVmL3Byb3BlcnRpZXMvYmFyIn19fSwidWlTY2hlbWEiOnt9fQ==)).

#### Expected behavior

Form to be rendered.

#### Actual behavior

No form rendered.  Console shows:

> **Error: Could not find a definition for #/definitions/testdef/properties/bar.**

### Version

0.46.0

### PR DESCRIPTION
**Pull request by eggyal**
### Reasons for making this change

Fixes #553

### Checklist

* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [x] I've updated docs if needed
  - [x] I've run `npm run cs-format` on my branch to conform my code to [prettier](https://github.com/prettier/prettier) coding style

### CODE REVIEW

**Comment by glasserc** [comment]
Thanks! The code looks good, but the code format check failed, for reasons that aren't related to your change. Maybe this is because we just updated to `prettier 1.2.2`. Could you rebase onto current master?

File: test/Form_test.js [hunk 1]
```diff
@@ -229,6 +229,27 @@ describe("Form", () => {
...
+          testdef: {
+            type: "object",
+            properties: {
+              bar: { type: "string" },
+            },
+          },
+        },
+        type: "object",
+        properties: {
+          foo: { $ref: "#/definitions/testdef/properties/bar" },
```
Comment (n1k0) [inline test/Form_test.js]: Actually I couldn't find any reference in the spec about this capability. The only part of it using nested definitions is [this one](http://json-schema.org/latest/json-schema-core.html#rfc.section.8.2) but I'm not sure about accessing `properties` like your implemention does would actually comply.

Do you by chance have a link describing how implementations should access nested definitions?

Comment (glasserc) [inline test/Form_test.js]: I couldn't find anything in the spec either, but I tried the given example on https://jsonschemalint.com/ and it works the way you would expect it to.

[Section 8.2.1](http://json-schema.org/latest/json-schema-core.html#rfc.section.8.2.1) says "Schemas can be identified by any URI that has been given to them, including a [JSON Pointer](https://tools.ietf.org/html/rfc6901)".

Comment (n1k0) [inline test/Form_test.js]: Ah, indeed I could validate the example in #553 as well, I think we're good to go then (well once the prettier formatting issue is addressed).

File: src/utils.js [hunk 2]
```diff
@@ -314,10 +314,21 @@ export function optionsList(schema) {
 
 function findSchemaDefinition($ref, definitions = {}) {
   // Extract and use the referenced definition if we have it.
-  const match = /#\/definitions\/(.*)$/.exec($ref);
-  if (match && match[1] && definitions.hasOwnProperty(match[1])) {
-    return definitions[match[1]];
+  const match = /^#\/definitions\/(.*)$/.exec($ref);
+  if (match && match[1]) {
+    const parts = match[1].split("/");
```
Comment (n1k0) [inline src/utils.js]: Makes me think of properties containing a `/` in their name, which will fail here. I don't want us to over this use case, but it may be a good idea to document this limitation, or at least adding a comment about it here.

Comment (eggyal) [inline src/utils.js]: @n1k0: Looking at the [JSON Pointer spec, §3. Syntax](https://tools.ietf.org/html/rfc6901#section-3), names containing `/` or `~` will be escaped as `~1` and `~0` respectively.  I will update to handle this, but I can't help wondering whether this project might instead use something like [json-schema-ref-parser](https://github.com/BigstickCarpet/json-schema-ref-parser) to handle these and other issues?

Comment (glasserc) [inline src/utils.js]: I'd be in favor of that! But if you don't have time, then just a FIXME comment would be fine.

Comment (n1k0) [inline src/utils.js]: I'd be in favor of it too, though their API being async it may be more work than needed in this specific case. A comment will be fine for now.

Also, please add a test case for this new covered edge case.

Comment (eggyal) [inline src/utils.js]: @n1k0: Test case added.

**Comment by n1k0** [comment]
Thank you :+1:

**Comment by n1k0** [comment]
Released in [v0.48.0](https://github.com/mozilla-services/react-jsonschema-form/releases/tag/v0.48.0).

File: src/utils.js (hunk 2 above)
Comment (eggyal) [reply src/utils.js]: @epicfaace: Perhaps I had misunderstood what @glasserc was proposing a FIXME comment for—I assumed it was to handle `/`s, which I fixed in 95bee41 so comment no longer required?

Comment (epicfaace) [reply src/utils.js]: > @epicfaace: Perhaps I had misunderstood what @glasserc was proposing a FIXME comment for—I assumed it was to handle `/`s, which I fixed in [95bee41](https://github.com/mozilla-services/react-jsonschema-form/commit/95bee418412810e67fb18c8d1a6c59c16d900ff6) so comment no longer required?

Oh, I assumed that the FIXME comment was supposed to be for using "json-schema-ref-parser to handle ... other issues."

Since you did fix the escaping of `/` and `~`, do you know if there's anything we're missing in this implementation which json-schema-ref-parser does well?
